File: native-ios/MGLFleetSDK/Sources/MGLFleetSDK/ReactParityBanner.swift
import Foundation

/// User-visible banners aligned with `app/page.tsx` (`bannerFromThrownError`, `bannerForOtpFailure`, …).
enum ReactParityBanner {
    static let validMobileError = "Please enter a valid mobile number"

    static let invalidFleetpayQr = "Invalid Fleetpay QR. Point at the station QR (must include tid)."
    static let selectVehicleFirst = "Select a vehicle first."
    static let noActiveFleet = "No active fleet — use your invite code."
    static let newUserContinueInvite = "New user — continue with invite code."
    static let inviteVerifyMobileFirst = "Invite flow incomplete — verify mobile first."
    static let useSendOtpOnLogin = "Use “Send OTP” on the login screen."
    static let completeMobileOtpFirst = "Complete mobile OTP verification first."
    static let sessionMissingInvite = "Session missing — go back to invite step."
    static let pinsDontMatch1f = "PINs don't match. Try again."
    static let pinsDidntMatchConfirm = "PINs didn't match. Try again."

    private static let bannerMsgMax = 280

    private static func errTxt(_ e: Error) -> String {
        var collected: [String] = []
        var current: Error? = e
        var depth = 0
        while let err = current, depth < 8 {
            depth += 1
            let localized = (err as? LocalizedError)?.errorDescription?
                .trimmingCharacters(in: .whitespacesAndNewlines)
            let fromNs = (err as NSError).localizedDescription
                .trimmingCharacters(in: .whitespacesAndNewlines)
            let piece: String
            if let l = localized, !l.isEmpty {
                piece = l
            } else if !fromNs.isEmpty {
                piece = fromNs
            } else {
                piece = ""
            }
            if !piece.isEmpty {
                collected.append(piece)
            }
            current = (err as NSError).userInfo[NSUnderlyingErrorKey] as? Error
        }
        if let last = collected.last { return last }
        return String(describing: type(of: e))
    }

    private static func isLikelyNetworkFailure(_ msg: String) -> Bool {
        let l = msg.lowercased()
        return l.range(of: "network|fetch|failed to fetch|timeout|502|503|504|econn|aborted", options: .regularExpression) != nil
    }

    private static let otpVerifyBannerFallback = "Could not verify OTP. Check your connection and try again."
    private static let pinVerifyBannerFallback = "Could not verify PIN. Check your connection and try again."

    private static let otpAuthNoisePattern =
        "unauthori[sz]ed|invalid[_\\s-]?grant|invalid[_\\s-]?token|\\b401\\b|bad\\s*credentials|access\\s*denied|^oauth\\s+failed|^http\\s*401|authentication\\s*failed|full\\s*authentication"

    private static let pinAuthNoisePattern =
        "unauthori[sz]ed|invalid[_\\s-]?grant|\\b401\\b|bad\\s*credentials|access\\s*denied|^oauth\\s+failed|^http\\s*401|wrong\\s*pin|incorrect\\s*pin|invalid\\s*pin|authentication\\s*failed"

    private static func humanizeOtpVerifyBanner(_ base: String, transportFallback: String) -> String {
        if base == transportFallback { return base }
        let t = base.trimmingCharacters(in: .whitespacesAndNewlines)
        if t.isEmpty || t == "Request failed." { return "Incorrect OTP. Try again." }
        if t.range(of: otpAuthNoisePattern, options: [.regularExpression, .caseInsensitive]) != nil {
            return "Incorrect OTP. Try again."
        }
        return base
    }

    private static func humanizePinVerifyBanner(_ base: String, transportFallback: String) -> String {
        if base == transportFallback { return base }
        let t = base.trimmingCharacters(in: .whitespacesAndNewlines)
        if t.isEmpty || t == "Request failed." { return "Incorrect PIN. Try again." }
        if t.range(of: pinAuthNoisePattern, options: [.regularExpression, .caseInsensitive]) != nil {
            return "Incorrect PIN. Try again."
        }
        return base
    }

    static func fromThrown(_ e: Error?, transportFallback: String) -> String {
        guard let e else { return transportFallback }
        let m = errTxt(e)
        if isLikelyNetworkFailure(m) { return transportFallback }
        let t = m.trimmingCharacters(in: .whitespacesAndNewlines)
        if t.isEmpty { return "Request failed." }
        if t.count <= bannerMsgMax { return t }
        let idx = t.index(t.startIndex, offsetBy: bannerMsgMax - 1)
        return String(t[..<idx]) + "…"
    }

    static func forOtpFailure(_ e: Error?) -> String {
        humanizeOtpVerifyBanner(
            fromThrown(e, transportFallback: otpVerifyBannerFallback),
            transportFallback: otpVerifyBannerFallback,
        )
    }

    static func forPinFailure(_ e: Error?) -> String {
        humanizePinVerifyBanner(
            fromThrown(e, transportFallback: pinVerifyBannerFallback),
            transportFallback: pinVerifyBannerFallback,
        )
    }

    static func forGenericFailure(_ e: Error?) -> String {
        fromThrown(e, transportFallback: "Something went wrong. Check your connection and try again.")
    }
}

File: native-android/fleet-sdk/src/main/java/com/mgl/fleet/sdk/demo/ReactParityBanner.kt
package com.mgl.fleet.sdk.demo

/**
 * User-visible API / validation banners aligned with [app/page.tsx]
 * (`bannerFromThrownError`, `bannerForOtpFailure`, `bannerForPinFailure`, `bannerForGenericFailure`).
 */
object ReactParityBanner {
    const val VALID_MOBILE_ERROR = "Please enter a valid mobile number"

    const val INVALID_FLEETPAY_QR = "Invalid Fleetpay QR. Point at the station QR (must include tid)."
    const val SELECT_VEHICLE_FIRST = "Select a vehicle first."
    const val LOGIN_FAILED_NO_TOKEN = "Login failed: no token."
    const val NO_ACTIVE_FLEET = "No active fleet — use your invite code."
    const val NEW_USER_CONTINUE_INVITE = "New user — continue with invite code."
    const val INVITE_VERIFY_MOBILE_FIRST = "Invite flow incomplete — verify mobile first."
    const val USE_SEND_OTP_ON_LOGIN = "Use “Send OTP” on the login screen."
    const val COMPLETE_MOBILE_OTP_FIRST = "Complete mobile OTP verification first."
    const val SESSION_MISSING_INVITE = "Session missing — go back to invite step."
    const val PINS_DONT_MATCH_1F = "PINs don't match. Try again."
    const val PINS_DONT_MATCH_CONFIRM = "PINs didn't match. Try again."
    const val COULD_NOT_REFRESH_TXN = "Could not refresh transactions (missing vehicle id)."

    private const val BANNER_MSG_MAX = 280

    private fun errTxt(t: Throwable): String {
        val chain = generateSequence(t) { it.cause }.toList().asReversed()
        for (ex in chain) {
            val m = ex.message?.trim().orEmpty()
            if (m.isNotEmpty()) return m
        }
        return t.javaClass.simpleName
    }

    private fun isLikelyNetworkFailure(msg: String): Boolean =
        Regex("network|fetch|failed to fetch|timeout|502|503|504|econn|aborted", RegexOption.IGNORE_CASE)
            .containsMatchIn(msg)

    private val otpVerifyBannerFallback = "Could not verify OTP. Check your connection and try again."
    private val pinVerifyBannerFallback = "Could not verify PIN. Check your connection and try again."

    private val otpAuthNoise =
        Regex(
            "unauthori[sz]ed|invalid[_\\s-]?grant|invalid[_\\s-]?token|\\b401\\b|bad\\s*credentials|access\\s*" +
                "denied|^oauth\\s+failed|^http\\s*401|authentication\\s*failed|full\\s*authentication",
            RegexOption.IGNORE_CASE,
        )

    private val pinAuthNoise =
        Regex(
            "unauthori[sz]ed|invalid[_\\s-]?grant|\\b401\\b|bad\\s*credentials|access\\s*denied|^oauth\\s+failed|" +
                "^http\\s*401|wrong\\s*pin|incorrect\\s*pin|invalid\\s*pin|authentication\\s*failed",
            RegexOption.IGNORE_CASE,
        )

    private fun humanizeOtpVerifyBanner(
        base: String,
        transportFallback: String,
    ): String {
        if (base == transportFallback) return base
        val t = base.trim()
        if (t.isEmpty() || t == "Request failed.") return "Incorrect OTP. Try again."
        return if (otpAuthNoise.containsMatchIn(t)) "Incorrect OTP. Try again." else base
    }

    private fun humanizePinVerifyBanner(
        base: String,
        transportFallback: String,
    ): String {
        if (base == transportFallback) return base
        val t = base.trim()
        if (t.isEmpty() || t == "Request failed.") return "Incorrect PIN. Try again."
        return if (pinAuthNoise.containsMatchIn(t)) "Incorrect PIN. Try again." else base
    }

    fun fromThrowable(
        e: Throwable?,
        transportFallback: String,
    ): String {
        val m = e?.let { errTxt(it) }.orEmpty()
        if (isLikelyNetworkFailure(m)) return transportFallback
        val t = m.trim()
        if (t.isEmpty()) return "Request failed."
        return if (t.length <= BANNER_MSG_MAX) t else "${t.take(BANNER_MSG_MAX - 1)}…"
    }

    fun forOtpFailure(e: Throwable?) =
        humanizeOtpVerifyBanner(fromThrowable(e, otpVerifyBannerFallback), otpVerifyBannerFallback)

    fun forPinFailure(e: Throwable?) =
        humanizePinVerifyBanner(fromThrowable(e, pinVerifyBannerFallback), pinVerifyBannerFallback)

    fun forGenericFailure(e: Throwable?) =
        fromThrowable(e, "Something went wrong. Check your connection and try again.")
}
